esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6118,7 +6104,259 @@ bool CWsDfuEx::onDFUFileAccess(IEspContext &context, IEspDFUFileAccessRequest &r
+    }
+    if (!expr)
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in parsing ECL: %s.", recordDefinition);
+    
+    if (!exportBinaryType(layoutBin, expr, false))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Failed in exportBinaryType.");
+}
+
+IFileDescriptor *CWsDfuEx::createDFUDescriptor(IEspContext &context, const char *logicalName,
+    const char *clusterName, const char *jobId, const char *userId, IUserDescriptor *userDesc)
+{
+    StringBuffer groupName;
+    getNodeGroupNameByClusterName(clusterName, groupName);
+    if (groupName.isEmpty())
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Cluster not found for %s.", clusterName);
+
+    StringBuffer basedir;
+    GroupType groupType;
+    Owned<IGroup> group = queryNamedGroupStore().lookup(groupName.str(), basedir, groupType);
+    if (!group)
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "No node group found for %s.", groupName.str());
+    if (groupType == grp_roxie)
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "Roxie cluster is not allowed: %s.", clusterName);
+
+    Owned<IFileDescriptor> fileDesc = createFileDescriptor();
+    fileDesc->queryProperties().setProp("@job", jobId);
+    if (!isEmptyString(userId))
+        fileDesc->queryProperties().setProp("@owner", userId);
+    StringBuffer curDir, defaultDir;
+    if (!getConfigurationDirectory(nullptr, "data", groupType == grp_hthor ? "hthor" : "thor", groupName, defaultDir))
+        makePhysicalPartName(logicalName, 0, 0, curDir, false, DFD_OSdefault); // legacy
+    else
+        makePhysicalPartName(logicalName, 0, 0, curDir, false, SepCharBaseOs(getPathSepChar(defaultDir)), defaultDir.str());
+    fileDesc->setDefaultDir(curDir.str());
+
+    StringBuffer partMask;
+    unsigned parts = group->ordinality();
+    getPartMask(partMask, logicalName, parts);
+    fileDesc->setNumParts(parts);
+    fileDesc->setPartMask(partMask);
+
+    ClusterPartDiskMapSpec mspec;
+    mspec.defaultCopies = DFD_DefaultCopies;
+    fileDesc->addCluster(groupName, group, mspec);
+
+    return fileDesc.getClear();
+}
+
+void CWsDfuEx::getFileAccessBeforePublish(IEspContext &context, const char *fileName, const char *jobId,
+    MemoryBuffer& layoutBin, Owned<IFileDescriptor> &fileDesc, IUserDescriptor *udesc, SecAccessFlags accessType,
+    IConstDFUFileAccessRequestBase &req, IEspDFUFileAccessInfo &accessInfo)
+{
+    IArrayOf<IEspDFUFilePart> dfuParts;
+    getFilePartsInfo(context, fileDesc, dfuParts);
+    accessInfo.setNumParts(fileDesc->numParts());
+    accessInfo.setFileParts(dfuParts);
+
+    StringBuffer metaInfo, expiryTime;
+    getFileMeta(metaInfo, fileName, fileDesc, udesc, expiryTime, jobId, req);
+    accessInfo.setMetaInfoBlob(metaInfo);
+    accessInfo.setExpiryTime(expiryTime.str());
+
+    if (layoutBin.length() == 0)
```

Comment:
if (layoutBin.length() == 0 && (req.getReturnJsonTypeInfo() || req.getReturnBinTypeInfo() )  shouldn't we generate an error, or at least a warn?

Reply:
I will add a warning to DFUFileCreateResponse.